### PR TITLE
Use compiler built ins

### LINKED ISSUE #370: Support getting default defines and include paths from the used compiler
Currently, when using ECC with CMake or a `compile_commands.json`, the used flags that are passed to the compiler include only the *explicitly set ones that are passed via the command line*. However, usually compilers also have some built-in include paths and defines that are set regardless of what flags are passed to the compiler by the user.

I guess when developing "host" applications, this usually is not a big problem, because the *usual* defines one would use are mostly the same across different compilers. However, especially when working with toolchains for embedded platforms, the target compiler often sets some special defines which then leads to ECC not being able to provide accurate completions in some cases.

Long story short:
It is possible to retrieve the defines that are set by default like this (also works for LLVM based compilers):

```bash
g++ -dM -E -x c++ -std=c++98 /dev/null
```

And the default include paths can be found by invoking a compiler like this (again this also works with LLVM based compilers):

```bash
g++ -v -x c++ -E /dev/null
```

It would be cool if ECC could query these defaults and append them explicitly to the flags passed to clang to get the auto-completions (of course, that should be optional so in case a compiler is used which does not allow this kind of retrieval or where the returned flags conflict with the ones clang uses internally, ECC remains functional).

As last time, I would volunteer to provide an implementation of this if this kind of feature is considered okay ;-)

### PR DESCRIPTION
**Pull request by mhoeher**
This PR tries to utilize compiler built-in flags for auto-completion. The idea is to query the compiler for built-in `#define`s and header search paths and append these flags in addition to other flags e.g. extracted from a `compile_commands.json` file for code completion. That way, if the user has a *special* compiler e.g for an embedded platform, `clang` is able to see platform specific defines.

This issue closes #370. In addition, it is related to #196 as - at least when using some of the completion methods - it allows the described "guessing" mechanism.

---

# Requirements #
- [x] Reference an issue in this PR. Create new one if needed.
- [x] Make sure your branch is based on `dev`.

### CODE REVIEW

File: EasyClangComplete.sublime-settings [hunk 1]
```diff
@@ -152,4 +152,13 @@
...
   ],
+
+  // Controls if we try to retrieve built-in flags from the compiler.
+  // Compilers tend to set some pre-defined symbols as well as include
+  // paths. If this option is on, we will ask the compiler for these and
+  // explicitly append them to the list of flags passed to clang when
+  // retrieving auto-completions.
+  // Generally, this option should improve the completion quality, however,
+  // if you encounter issues, you can turn it off.
+  "use_compiler_built_in_flags": false,
```
Comment (niosus) [inline EasyClangComplete.sublime-settings]: I feel like we can enable it by default, removing the default values for `common_flags` as they would be guessed from the compiler itself.

File: plugin/flags_sources/cmake_file.py [hunk 2]
```diff
@@ -42,7 +42,8 @@ def __init__(self,
                  prefix_paths,
                  flags,
                  cmake_binary="cmake",
-                 header_to_source_mapping=None):
+                 header_to_source_mapping=None,
+                 use_compiler_builtins=False):
```
Comment (niosus) [inline plugin/flags_sources/cmake_file.py]: Let's probably remove the default value. This will ensure that we are always propagating the settings correctly. I will then add a separate PR to remove other default values here. They should not be there. It just adds a new point of breakage. :)

File: plugin/flags_sources/compilation_db.py [hunk 3]
```diff
@@ -145,6 +149,24 @@ def _parse_database(self, database_file):
...
+                # combination, the results are cached by the class internally.
+                builtins = CompilerBuiltIns(argument_list)
+
+                # Append built-in flags to the end of the list:
+                # Note: We keep the last argument as last, as it
+                # usually is the file name.
+                argument_list = (
+                    argument_list[:-1] + builtins.flags +
+                    argument_list[-1:])
+                #argument_list += builtins.defines
```
Comment (niosus) [inline plugin/flags_sources/compilation_db.py]: nit: please remove unused code.

File: plugin/flags_sources/compilation_db.py [hunk 4]
```diff
@@ -145,6 +149,24 @@ def _parse_database(self, database_file):
                 log.critical(" compilation database has unsupported format")
                 return None
 
+            # If enabled, try to retrieve default flags for the compiler
+            # and language combination:
+            if self._use_compiler_builtins:
+
```
Comment (niosus) [inline plugin/flags_sources/compilation_db.py]: nit: please remove empty line

File: plugin/utils/compiler_builtins.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
+"""Get compiler built-in flags."""
+
+import logging as _logging
+
+_log = _logging.getLogger("ECC")
```
Comment (niosus) [inline plugin/utils/compiler_builtins.py]: nit: maybe I lack some knowledge here. Why would we start the name with underscore? This is not a protected class member, right?

File: plugin/utils/compiler_builtins.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+            compiler = args[0]
+        else:
+            self._debug("Got empty command line - cannot extract compiler")
+        if len(args) > 1:
+            for arg in args[1:]:
+                if arg.startswith("-std="):
+                    std = arg[5:]
+        return (compiler, std)
+
+    def _guess_language(self, compiler):
```
Comment (niosus) [inline plugin/utils/compiler_builtins.py]: I have two questions to this.
- This will not work with objective c, right? We should make it clear in the settings, I guess. I am not the one who added the support for it in the first place, but I don't want to break it.
- Can we unify this function with [this](https://github.com/niosus/EasyClangComplete/blob/master/plugin/tools.py#L522) function?

File: plugin/utils/compiler_builtins.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+        result = list()
+
+        args = [compiler, '-x', language]
+        if std is not None:
+            args += ['-std=' + std]
+        args += ['-Wp,-v', '-E', '-']
+
+        try:
+            res = subprocess.Popen(
```
Comment (niosus) [inline plugin/utils/compiler_builtins.py]: Do you think we can reuse a [function from tools](https://github.com/niosus/EasyClangComplete/blob/master/plugin/tools.py#L687)?

File: plugin/utils/compiler_builtins.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+            args = split(args)
+        # Guess the compiler and standard:
+        (compiler, std) = self._guess_compiler(args)
+        if compiler is not None:
+            # Guess the language (we need to pass it to the compiler
+            # explicitly):
+            language = self._guess_language(compiler)
+            # Get defines and include paths from the compier:
+            cfg = (compiler, std, language)
+            self._debug("Getting default flags for {}".format(cfg))
```
Comment (niosus) [inline plugin/utils/compiler_builtins.py]: Use log here?

File: plugin/utils/compiler_builtins.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+            if cfg in CompilerBuiltIns.__cache:
+                self._debug("Reusing flags from cache")
```
Comment (niosus) [inline plugin/utils/compiler_builtins.py]: Same as above.

File: plugin/utils/compiler_builtins.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        # Remove some default flags which get set by clang itself.
+        # Otherwise, we will get error later:
+        for flag in CompilerBuiltIns.__DEFINES_BLACKLIST:
+            prefix = "-D%s" % flag
+            for define in defines:
+                if define.startswith(prefix):
+                    defines.remove(define)
+        return defines
+
+    def _debug(self, msg):
```
Comment (niosus) [inline plugin/utils/compiler_builtins.py]: There seems to be no need for these functions from my point of view. The name of the class and function calling the log is already part of logging module.

File: plugin/utils/compiler_builtins.py [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+        _log.debug("[compilerbuiltins] %s" % msg)
+
+    def _warn(self, msg):
```
Comment (niosus) [inline plugin/utils/compiler_builtins.py]: Same as above. I would prefer to just use logging.

**Comment by niosus** [comment]
@mhoeher in https://github.com/niosus/EasyClangComplete/issues/370#issuecomment-352697411 you're saying that this code is only in use when using cmake or the compilation database. I feel like we should actually switch to using this code by default, replacing the placeholder like `common_flags` with this. Do you see anything that can stop us from this?

**Comment by mhoeher** [comment]
@niosus I think we can definitely try to push to make use of the built-in flags the default. However, I think we should be careful entirely replacing the existing flags mechanism: During development I already stumbled upon 4 defines which we probably must not set explicitly as otherwise clang will - rightfully - complain. I can quite well imagine that this might be true for other defines or even includes with different tool chains. My conclusion would hence be that:

1. We should try to make guessing the right flags from the compiler (if available) the default, but:
2. Keep the existing solution at least as a fallback in case clang code completion breaks with the built in flags.

Regarding the scoped use of the feature: For it to work properly we must know the target compiler. If I understand the existing flags sources implementations correctly, this is only the case with `cmake` or a `compile_commands.json` as source. Using the `.clang-complete` approach, we anyway use `clang` as the compiler and hence, there is no need to explicitly set these flags from the outside (as `clang` implies them on itself). Please correct me if I'm wrong here ;-)

**Comment by niosus** [comment]
Maybe you're right. However, I was expecting that we would always use the compiler specified in `clang_binary` setting. We should also probably use it for generating compilation database with cmake. This way it is per-project configurable and cleanly defaults to clang.

File: plugin/utils/compiler_builtins.py (hunk 5 above)
Comment (mhoeher) [reply plugin/utils/compiler_builtins.py]: The underscore notation is actually not only used inside classes but also for members inside modules. Background is, that you make clear - in analogy to a class definition - which parts of your module are belonging to the interface and which are internal and hence shall not be used. This is required due to the way Python works. After doing a

```python
import logging
```

in a file called `my_mod.py`, we can

```python
from my_mod import logging
```

which obviously is nonsense, but works as you would expect.

Long story short: I think we can drop the underscore notation (as this is not a library which is supposed to be used in other apps or libraries) but it's just a bit cleaner. I don't mind too much, so its your call here ;-)

File: plugin/utils/compiler_builtins.py (hunk 8 above)
Comment (mhoeher) [reply plugin/utils/compiler_builtins.py]: I added a `_debug` utility method which writes to the log (and adds a prefix so we quickly see the message originates from that class). Is this what you mean?

File: plugin/utils/compiler_builtins.py (hunk 6 above)
Comment (mhoeher) [reply plugin/utils/compiler_builtins.py]: Right, currently, this only distinguishes C++ from the "rest" (where we assume that rest refers to C). I think we can try to pimp this utility a bit to also make it work with other languages, I just never works with Objective C/C++ before and hence did not care in this first shot yet.

Regarding the existing method you mentioned: I already stumbled upon it, however, I chose not to make use of it as it requires a `view` object and I don't see how to get in in the place where I do the guessing thing. If you see a way how we can properly pass it through, please let me know and I'll try to implement it.

File: plugin/utils/compiler_builtins.py (hunk 5 above)
Comment (niosus) [reply plugin/utils/compiler_builtins.py]: Thanks for explaining this. Today I learned smth about python. I am completely convinced.

File: plugin/utils/compiler_builtins.py (hunk 7 above)
Comment (mhoeher) [reply plugin/utils/compiler_builtins.py]: Ah, cool, definitely, I think so! One question: In the existing code I catch the potential `FileNotFoundError`; if I see it correctly, this is not done in the `run_command` utility. Shall we catch that exception in `run_command` or further up in the caller?

File: plugin/utils/compiler_builtins.py (hunk 10 above)
Comment (mhoeher) [reply plugin/utils/compiler_builtins.py]: Okay, I just had in mind that in other files you used to have dedicated prefixes in the log messages, so I wanted to keep that aligned. If this is not required, I will happily convert to plain logging ;-)

File: plugin/utils/compiler_builtins.py (hunk 6 above)
Comment (niosus) [reply plugin/utils/compiler_builtins.py]: I did not work with objective c either, but we should at least make it clear to the user that it *is not supposed* to work as opposed to just silently failing or generating garbage. I guess a clean fallback to using `common_flags` would be appropriate if the language is different from c or c++. This way if somebody decides to use it with objective c we can ask them to adapt the code to account for the missing functionality :)

Comment (niosus) [reply plugin/utils/compiler_builtins.py]: As for the function and what it does. Currently, I feel like this is not a very scalable solution. For example, `clang++` actually symlinks to `clang`: 
```
$ ls -la | grep clang             
-rwxr-xr-x 1 root root 22938632 Jul 12  2016 clang
lrwxrwxrwx 1 root root        5 Jul 12  2016 clang++ -> clang
```
Am I correct that if the user would use `clang` this would report `c`?

**Comment by mhoeher** [comment]
Not always. When you anyhow use a `compile_commands.json`, the compiler used is actually part of the flags:

```json
[
{
  "arguments": [
    "/usr/bin/c++",
    "-I/lib_include_dir",
    "-o",
    "CMakeFiles/main_obj.o",
    "-c",
    "/home/user/dummy_main.cpp"
    ],
  "directory": "/main_dir",
  "file": "/home/user/dummy_main.cpp"
},
"..."
```
Hence, when a compile commands is used as source (which indirectly also is the case when using `cmake`) we take the information from this file instead. The currently only remaining flag source where this is not the case is when using a flags file, where we anyhow call into clang, so for the moment I saw no need to hook the new functionality in there as well.

Is it possible (and supported) to specify another (non-clang) compiler via `clang_binary`? If so, my assumption would definitely be void and we should add the built ins.

Otherwise, I would actually not try to use `clang_binary` for the other cases but instead always prefer what we get from the underlying build system, because it usually "knows" better (at least that would be my assumption).

File: plugin/utils/compiler_builtins.py (hunk 6 above)
Comment (mhoeher) [reply plugin/utils/compiler_builtins.py]: Regarding your first comment: Yes, the function definitely needs some love - silently falling back to C should probably not be used by default.

Regarding your second comment: It is normal that `clang++` is sym-linked to `clang` - this is a normal approach also for other compilers. The `clang` binary internally checks by which name it was called and either activates C or C++ mode (at least this is my understanding).

The background the method is written like it that is, that it is called on a compiler command line like

```
path/to/c++ -I/some/include/path -DSOME_DEFINE my_file.cpp
```
To get the correct defines and include paths for code completion, we must - when asking the compiler - explicitly tell it for which language, i.e. we must pass it the `-x SOMELANGUAGE` flag - "just" using the correctly symlinked version will for whatever reasons not work as expected :-(

Currently, the method tries to derive the language only from the compiler name. This is not quite smart (e.g. if the command line in a `compile_commands.json` contains a `-x` flag, we should probably use that instead of looking at the compiler name), but it works for the beginning.

I think your concerns about what a user might select would only kick in if we do not work on a given call to a compiler (as we get it from CMake and a compile database) but instead would have to guess from other pieces of information. In these cases, e.g. the selected language of the current view or the extension of a file should rather be used to decide.

Comment (mhoeher) [reply plugin/utils/compiler_builtins.py]: I just looked into the code once more and also into how e.g. a typical `compile_commands.json` generated by CMake looks like. My observation for now is:

* By default, CMake does not include a `-x` flag from which we can guess the language from.
* On my system, I checked the `gcc-objc` and `gcc-objc++` packages: They do *not* seem to install additional binaries; so I guess if one would like to use them CMake and other tools would have to explicitly ask for a language using the `-x` flag.
* In addition, I compared the list of flags that we get e.g. from GCC when asking for different languages (C, C++ and ObjC). The lists seems vastly identical, with some specific defines usually added to indicate a particular language's capabilities. Hence, I would assume that the flags for C are mostly a subset of all other supported languages.

Hence, I would indeed suggest the following:

1. When guessing the language, we check first if we find an explicit `-x` flag and use it regardless what else is in the command line.
2. In any other case, we try to guess the name from the compiler. If the compiler name ends with `c++` or `++`, we assume the file in question is compiled with C++, otherwise, we assume C. That should be relatively save, because the flags used for C form (mostly) a subset of any other languages' flags.

If we would not allow point 2., we probably could not provide this feature for C sources altogether (at least I currently don't see how we could).

What do you think?

**Comment by niosus** [comment]
Oh, this feels hackier and hackier by the minute. :) 

Alright. Let's try to break this up.

All the completion is happening with clang, so the code *must* be built with clang. This causes problems as there are different tool chains that are designed to run with a different compiler. So we should get the needed flags from that respective compiler, right?

We have a couple of cases:
- [ ] `common_flags` setting. Does not specify a compiler on its own. Use clang from `clang_binary` setting here? So this method **knows nothing** about the actual compiler.
- [ ] `.clang_complete` file. Starts feeling obsolete at this point as it just duplicates the `common_flags`. Also **knows nothing** about the actual compiler.
- [ ] `CMake`. This method generates a compilation database. Refer to next entry for details. But if we  force it to use clang as compiler then it **does not know** about the real compiler.
- [x] Compilation database. Implicitly defines a compiler by specifying it as the first argument of every command.

I have marked with the check boxes if the tool knows about the actual compiler.

It seems that the compilation database generated elsewhere is the only one who knows what we are supposed to use to get the flags. I feel that at this point it would make sense to change your setting to smth like `"target_compiler"` that will be `"none"` by default and will define a compiler to use for getting proper flags. We should think how to handle the case with overriding this compiler with the compilation database though. 

How does this sound. Did I miss anything?

File: plugin/utils/compiler_builtins.py (hunk 6 above)
Comment (niosus) [reply plugin/utils/compiler_builtins.py]: I think this is fine for beginning. I am still thinking more in the direction of deriving the correct language from a view, even if it means that we need to propagate it down into this class (or just get language beforewise). The good part about using the view-dependent detection is that there is a closed loop to the user. I mean, the user sees which syntax is being used for a file, which makes it easy for us to avoid ambiguities. 

But again, I believe your current suggestion is a good first step. Just probably leave a TODO in the code to not forget that there can be an alternative.

File: plugin/utils/compiler_builtins.py (hunk 7 above)
Comment (niosus) [reply plugin/utils/compiler_builtins.py]: Yeah, it just never occured yet :) Let's catch it there to be consistent.

**Comment by mhoeher** [comment]
Personally, I would not call it hacky. It's maybe just not the magical answer to all problems code-completion ;-)

Your break down matches reality quite nicely. Of all the flag sources currently supported, only in case we use the compilation DB (either directly or via CMake) provides us the target compiler from which we could try to derive additional information to make code completion via clang better. Consequentially, the new functionality is baked exclusively into these flag source implementations at the moment.

When not using a compilation DB, we cannot know the target compiler and just assume we use clang. In this case, I think there is no value to integrate passing in built in flags as we wouldn't provide anything new to clang.

The only additional thing I could imagine is, that when not using a compilation DB, a user might specify the used compiler on a per-project level. In this case, we could query that compiler's built-ins when using the `common_flags` or `.clang_complete` approach.

Bottom line, I would suggest the following:

* We make sure the detection of compiler and language works reliably. This way we best serve use cases where a compilation DB is in use.
* To clarify stuff, I will change the code and settings to refer to the term `target_compiler` where appropriate to ensure users and contributors are aware that there might actually be two tools involved (one compiler they use in their build chain and one that is used to get code completion).
* We add a new `target_compiler` option (or maybe even better: One per language). This option would be used when no compilation DB is present. Furthermore, we could pass these compilers to CMake when it is called to generate the `compile_commands.json`.

Does this sound like a reasonable plan? As in my last PR, please keep in mind that design decisions on my side are heavily driven by the selfish need for good code completion for a project I am working on and might not necessarily be of interest for a greater audience otherwise ;-)
So, please let me know if things are going in the right direction.

**Comment by niosus** [comment]
Yeah, I have a pretty low tolerance towards using the word "hacky". It is not meant to offend in any way and just represents the amount of things one needs to remember to keep this machine running :)

> When not using a compilation DB, we cannot know the target compiler and just assume we use clang. In this case, I think there is no value to integrate passing in built in flags as we wouldn't provide anything new to clang.

There is a value to this as libclang and clang are not always using exactly the same settings.

Now with this out of the way, I generally agree with what you wrote and thanks for pointing out that we need multiple `target_compilers` for multiple languages.

I totally understand your motivation and it was my initial motivation from the start of this whole project :) So my goal is to squeeze the most of your implementation to have it as close as possible to the simplicity of support that I am striving for without you actually giving up on it :)

Overall, I think we agree on what it should look like and given, that result is modular enough, it should also be relatively easy for me to add whichever nitpicking comes to my mind later :) 

Would you also have the time for adding the tests for the new class? Otherwise, when we are done with this PR I can do this myself, which will also let me understand the code better, but will make the merging of the PR a bit slower.

**Comment by mhoeher** [comment]
No worries.I also am totally with you on keeping things as simple as possible ;-)

I will try to go ahead and implement the points as outlined above. I will also try to implement some unit tests before the final merge. I just have to think about how to do this in a proper way, because I am afraid these built-ins might differ from platform to platform a bit :-o

Regarding differences between libclang and clang: Thanks, I was not aware about this. That means that the last point I mentioned even makes more sense, so I'll definitely try to get this implemented as well.

A question back regarding that point: Does ECC currently support project specific settings? If I understand the settings manager class correctly, it does not, but maybe I fail to understand it correctly. Especially for the target compiler settings, project specific settings would definitely make sense.

**Comment by niosus** [comment]
> Regarding differences between libclang and clang
Yeah, that has caused quite some pain making stl work with libclang on all platforms :)

Cool! Feel free to ask anything or criticize anything you find yourself fighting again and thanks so much for doing this!

**Comment by mhoeher** [comment]
Hi @niosus,
I tried to incorporate more of your suggestions. In addition, I tried to come up with a unit test for the compiler built ins. Could you check if this test is sufficient for now? Focus when writing was on making sure that the guessed language is as accurate as possible.

Unfortunately, the style unit test currently fails on Travis (and also locally), however, I fail to see why. :-o Do you have an idea? I bet it is due to the changes in the `run_command()` I did, but I don't see how they could affect the test.

**Comment by niosus** [comment]
Ok, so the failure seems to be because of pep8 compliance. Run [similar](https://github.com/niosus/EasyClangComplete/blob/master/tests/test_style.py#L17) command locally to find out what it complains about.

I will have a look at the other stuff today or tomorrow. Thanks!

**Comment by mhoeher** [comment]
Ah, dammit. I though I've seen pep8 findings printed out before elsewhere when running the full tests so the idea that this test just executes code checks never come to my mind ^^' Should pass now.

Unless you find anything else, the only remaining point is to make use of explicitly set target compilers when not using a compilation DB.

**Comment by mhoeher** [comment]
Moin moin,
I think I am done so far.

* When a user set target compilers, their built in flags are now added on top to the common flags. This should improve completions for simple projects without any additional information that can be used.
* In addition, when using CMake as flags source, if a C or C++ compiler is set, it is passed via a toolchain file to CMake so it picks up the appropriate compiler.

I hope that covers most use cases ;-)

**Comment by niosus** [comment]
Cool! Thanks so much! Sorry for the delay. I will have a look somewhere in this holiday season, but cannot promise how quickly this will happen :) Merry Christmas :)

**Comment by mhoeher** [comment]
No hurry, take whatever time you need to do a proper review ;-)

Wish you too a nice holiday season without too much work (and code reviews ;P )!

**Comment by niosus** [comment]
Sorry for a huge delay. I have briefly looked into this and it looks ok. There is a small number of places that might need some improvement, but I don't want to close the cycle on me for this. So I decided to merge it into dev for now and give it a spin for the next couple of days. Hopefully, I will merge it to master and release a new version on the weekend.

**Comment by niosus** [comment]
@mhoeher just wanted to let you know that if I enable this setting it breaks my workflow and my code sometimes fails to compile. Probably clang cannot deal with some flags that are generated by gcc toolchain. I will keep investigating this. Maybe this should not be enabled by default in the end. Don't know yet.

**Comment by mhoeher** [comment]
@niosus it would be perfectly fine for me if the setting is optional. If required, we could also use a kind of "blacklist" for some defines - I also have the feeling that some of the defines should actually not be set, but had no time yet to create a working minimal example to prove this.

**Comment by niosus** [comment]
I feel like we're tapping into something really deep over here :) 
I believe I will indeed make it optional over the weekend and then we can research in a leisurely pace.

**Comment by niosus** [comment]
I also wonder if just disabling the underscored defines would do the trick.

**Comment by mhoeher** [comment]
I currently have no concrete case where I encounter an issue, hence I'd suggest:

1. Make the setting optional for the time being to avoid users running into trouble with it.
2. Then, we can try to find out how to improve the functionality.

The issue could exactly be with the underscore defines, but unfortunately, they are also used e.g. by special toolchains, so this would potentially render the built ins retrieval useless. Maybe, instead of doing a boolean option (use built ins yes/no?) the option could be changed to a mode string:

* "on" - retrieve built ins.
* "off" - do not retrieve built ins.
* "no_underscores" - retrieve built ins but omit ones starting with an underscore.
* "..." - whatever else could be useful.

Alternatively, we could add another option where the user could pass in e.g. a regular expression to filter out certain defines:

```json
{
    "omit_built_ins_matching": "^_+" 
}
```
This option could then be used to implement a default blacklist (if we can determine a set of "bad" defines we better don't include) which users could override if desired.

**Comment by niosus** [comment]
This sounds reasonable. Let me think about it a bit.
